ethcore/src/client/client.rs

```diff
@@ -1231,12 +1231,11 @@ impl BlockChainClient for Client {
 	}
 
 	fn estimate_gas(&self, t: &SignedTransaction, block: BlockId) -> Result<U256, CallError> {
-		const UPPER_CEILING: u64 = 1_000_000_000_000u64;
-		let (mut upper, env_info)  = {
+		let (upper, env_info)  = {
 			let mut env_info = self.env_info(block).ok_or(CallError::StatePruned)?;
-			let initial_upper = env_info.gas_limit;
-			env_info.gas_limit = UPPER_CEILING.into();
-			(initial_upper, env_info)
+			let upper = env_info.gas_limit * U256::from(100); // enough gas for estimate_gas
```

Comment:
The loggic is quite different now.

Previously:
1. Start with `block_gas_limit`
2. If that's not enough proceed with `UPPER_CEILING`.

Currently:
1. Start with `block_gas_limit * 100`

I think we should keep the starting point at `block_gas_limit`, since it will converge much faster for valid transactions and fallback to `100 * limit` only if the first fails.

Reply:
I think a execution time in estimate_gas of a valid transaction with `block_gas_limit` is a same with `block_gas_limit * 100`, therefore, I changed to one phase estimate_gas from two phases.

I compared a execution time with `block_gas_limit` and `block_gas_limit * 100`.
```
> var calcValid = function() {
... var start = new Date().getTime();
... sample.validTx.estimateGas(function (err, gas) {
..... var end = new Date().getTime();
..... console.log(end - start);
..... });
... }
```

The following result is a execution time with `block_gas_limit`.
```
> calcValid()
undefined
> 75
```

The following is with `block_gas_limit * 100`.
```
> calcValid()
undefined
> 80
```
These result shows a execution time in estimate_gas of a valid transaction with `block_gas_limit` is a same with `block_gas_limit * 100`.

Reply:
This time will depend on complexity of the transaction (CPU, IO). You can see that there is even a special case for transfer transactions (they will return immediately, by checking the lower bound first). I'm pretty positive that `block_gas_limit` should converge ~6 times faster.

Reply:
I understood.
I will change PR as you say.

By the way, I think `block_gas_limit * 100` is so large, but what do you think?

Reply:
It's pretty big, but afair users requested that estimate gas should work way over block gas limit. It's a significant improvement (in terms of time to fail) over what we have now, so I'm in favour of including this PR (or `block_gas_limit * 10`) for 1.9 and probably re-visiting it if we see that it still takes too much time for users.

Reply:
OK, thank you.
I fixed.